README.rst
django-sitetree
===============
http://github.com/idlesign/django-sitetree

.. image:: https://img.shields.io/pypi/v/django-sitetree.svg
    :target: https://pypi.python.org/pypi/django-sitetree

.. image:: https://img.shields.io/pypi/dm/django-sitetree.svg
    :target: https://pypi.python.org/pypi/django-sitetree

.. image:: https://img.shields.io/pypi/l/django-sitetree.svg
    :target: https://pypi.python.org/pypi/django-sitetree

.. image:: https://img.shields.io/coveralls/idlesign/django-sitetree/master.svg
    :target: https://coveralls.io/r/idlesign/django-sitetree

.. image:: https://img.shields.io/travis/idlesign/django-sitetree/master.svg
    :target: https://travis-ci.org/idlesign/django-sitetree

.. image:: https://landscape.io/github/idlesign/django-sitetree/master/landscape.svg?style=flat
   :target: https://landscape.io/github/idlesign/django-sitetree/master


What's that
-----------

*django-sitetree is a reusable application for Django, introducing site tree, menu and breadcrumbs navigation elements.*

Site structure in django-sitetree is described through Django admin interface in a so called site trees.
Every item of such a tree describes a page or a set of pages through the relation of URI or URL to human-friendly title. Eg. using site tree editor in Django admin::

  URI             Title
    /             - Site Root
    |_users/      - Site Users
      |_users/13/ - Definite User


Alas the example above makes a little sense if you have more than just a few users, that's why django-sitetree supports Django template tags in item titles and Django named URLs in item URIs.
If we define a named URL for user personal page in urls.py, for example, 'users-personal', we could change a scheme in the following way::

  URI                           Title
    /                           - Site Root
    |_users/                    - Site Users
      |_users-personal user.id  - User Called {{ user.first_name }}

After setting up site structure as a sitetree you should be able to use convenient and highly customizable site navigation means (menus, breadcrumbs and full site trees).
User access to certain sitetree items can be restricted to authenticated users or more accurately with the help of Django permissions system (Auth contrib package).


Documentation
-------------

http://django-sitetree.readthedocs.org/


CHANGELOG
django-sitetree changelog
=========================

v1.5.0
------
+ Added Norwegian translation.
+ Added SITETREE_RAISE_ITEMS_ERRORS_ON_DEBUG setting (see #157).
+ Exposed SITETREE_CACHE_TIMEOUT setting.
+ Added `as` clause support for `sitetree_page_title`, `sitetree_page_description` and `sitetree_page_hint` template tags.
* Fixed cache problems when using sitetree_resync_apps  (see #135, #166).
* Fixed disappearing tree items for guests in Admin contrib (Django 1.8) (see #164).
* Fix deprecation warning in Django 1.8 (see #178).
* Fixed permissions check for dynamic tree items (see #165).


v1.4.0
------
+ Introduced Django 1.8 support (see #152).
* Fixed extra spaces issue in breadcrumbs (closes #150).


v1.3.0
------
+ Implemented `django-smuggler` thirdparty support.
+ Implemented `i18n_patterns` compatibility (closes #148).
* Fixed menu alias clash with context variable (closes #117).


v1.2.1
-------
* `{{ STATIC_URL }}` is replaced with `{% static %}` admin templates.
* Fixed `sitetreeload` management command compatibility with py3.
* Fixed `sitetreeload` management command compatibility with Django 1.7.


v1.2.0
-------
+ Added support for both South and Django 1.7 migrations.
+ `register_dynamic_trees()` now accepts `reset_cache` kwarg.


v1.1.0
-------
+ Django 1.7 ready.
+ `item()` now accepts `access_by_perms` and `perms_mode_all`.
* Fixed `SiteTreeError` for `TEMPLATE_CONTEXT_PROCESSORS` when not `DEBUG`.
* Fixed `this-siblings` item alias behavior.
* `register_dynamic_trees()` refactored to a less-brackets style.
* `LazyTitle` objects made py3 compatible.
* Fixed "Save and continue editing" on Tree Item yields "does not exist" error.
* Global context is change is now tested with `id()`.


v1.0.0
-------
+ Added `breadcrumbs-title` template.
+ `UNRESOLVED_ITEM_MARKER` introduced to settings.
+ Added Django 1.6 test rules.
* Fixed django 1.4.10 issue (see #116).
- Dropped deprecated `sitetree_url` tag arguments.
- Dropped deprecated template var support in sitetree item URL field.


v0.12.1
-------
+ Fixed bug when running sitetree with django version 1.4.10


v0.12.0
-------
+ Implemented runtime defined (dynamic) trees (see #105).
+ Added Foundation breadcrumbs template.
+ Added Semantic UI templates.
* Fixed `DatabaseError: integer out of range` generated on `sitetree_resync_apps` (see #105).
* Sitetree for apps module name made adjustable and defaults to `sitetrees` to prevent module name clashes.


v0.11.0
-------
+ Implemented helpers for dynamic tree structuring (see #105)
+ Added experimental support for app-defined trees (see #105).
+ Implemented tree item access restriction "for guests only" (closes #108)
+ Added Bootstrap 3 templates (closes #100).


v0.10.0
-------
+ Added experimental support for user-defined sitetree models (see #64).
+ Implemented `sitetree_page_hint` template tag (closes #103).
+ Added Spanish translation (closes #101).


v0.9.5
------
+ Added `sitetree_page_description` template tag.
+ Tree item dropdown experimentally exposed (see #84).
+ Added `this-ancestor-siblings` alias support (see #99).
* Fixed 'map' object has no attribute 'append' on Py3 (see #94).
* Fixed sort order changing for top level tree items (see #98).


v0.9.4
------
+ Added Django 1.5 url tag syntax support for URL field.
* Fixed args quoting in url tag call causing #unresolved links (closes #95).


v0.9.3
------
+ Added Python 3 support (see #93).
+ Added persian translation.
* Fixed quotes around slug-like arguments in url() (#90).
* Added deprecation warnings workarounds (#89).
- Dropped Django 1.3 support.


v0.9.2
------
+ Django 1.3+ set as required minimum.
+ Added Django 1.5 support (see #87).
+ Added namespaces support for known urls hint (closes #83).
+ Added 'as' clause for sitetree_url tag (see #81).
+ Introduced Tree.get_title().
* Humble performance improvements for large trees (see #49).
* Template variables in URL feature marked deprecated (fixes #74).


v0.9.1
------
+ Added experimental URL pattern name hinting at tree item create/edit page (see #67).
* Fixed tree current item detection failures with i18n trees (see #66).


v0.9.0
------
+ Added support for tree and tree item admin models overriding (see #43).
+ Added Foundation CSS Framework menu basic templates (see #41).
+ Added Bootstrap CSS Framework menu and breadcrumbs basic templates (see #41).
* Fixed sorting tree items by parent ID instead of parent sort order (see #55).
* Fixed crashes when using sitetreeload command with colored terminal (fixes #61).


v0.8.0
------
+ Added 'title' field for Tree models.
+ Added German translation.
* Fixed management directory missing from dist (closes #50).


v0.7.0
------
+ Added 'sitetreedump' and 'sitetreeload' management commands (fixes #36).
+ Added tests runner 'runtests.py'. Unit tests are now self-contained (see #42).
+ Now 'sitetree_menu' tag accepts any string as target branch alias. Related to dashed strings (see #38).
+ Now 'has_children' tree item attribute stands for visible children (see #38).
+ Now global template context is passed down all sitetree-related templates (see #39, see #40).
* Fixed template tags params clashes with context variables (see #42).


v0.6.0
------
+ Added i18n trees support (see #27).
+ Added tree items processing hook support (see #26).
* Fixed various tree items filtering issues (see #34, #35).


v0.5.2
------
* Fix for yes/no icons not shown in tree admin when non-eng LANGUAGE_CODE is used (closes #31).
* Fixed item's depth calculation (closes #29).
* Fixed misleading tree rendering bug in admin interface.
* Improved Django 1.4 compatibility (fixes #33).


v0.5.1
------
+ Added Django 1.4 static files compatibility for Admin contrib.
+ Added support for slug-like IDs as item lookup params (see #24).

sitetree/admin.py
from django.conf import settings as django_settings
from django.core.urlresolvers import get_urlconf, get_resolver
from django.utils.translation import ugettext_lazy as _
from django.utils import six
from django.http import HttpResponseRedirect
from django.contrib import admin
from django.contrib.admin.sites import NotRegistered
from django.contrib import messages
from django.conf.urls import patterns, url

from .settings import MODEL_TREE, MODEL_TREE_ITEM
from .fields import TreeItemChoiceField
from .utils import get_tree_model, get_tree_item_model, get_app_n_model


SMUGGLER_INSTALLED = 'smuggler' in django_settings.INSTALLED_APPS

MODEL_TREE_CLASS = get_tree_model()
MODEL_TREE_ITEM_CLASS = get_tree_item_model()

_TREE_ADMIN = lambda: TreeAdmin
_ITEM_ADMIN = lambda: TreeItemAdmin


def get_model_url_name(model_nfo, page, with_namespace=False):
    """Returns a URL for a given Tree admin page type."""
    prefix = ''
    if with_namespace:
        prefix = 'admin:'
    return ('%s%s_%s' % (prefix, '%s_%s' % model_nfo, page)).lower()


def get_tree_url_name(page, with_namespace=False):
    """Returns a URL for a given Tree admin page type."""
    return get_model_url_name(get_app_n_model('MODEL_TREE'), page, with_namespace)


def get_tree_item_url_name(page, with_namespace=False):
    """Returns a URL for a given Tree Item admin page type."""
    return get_model_url_name(get_app_n_model('MODEL_TREE_ITEM'), page, with_namespace)


_TREE_URLS = {
    'app': get_app_n_model('MODEL_TREE')[0],
    'change': get_tree_url_name('change', 'admin:'),
    'changelist': get_tree_url_name('changelist', 'admin:'),
    'treeitem_change': get_tree_item_url_name('change', 'admin:')
}


def _reregister_tree_admin():
    """Forces unregistration of tree admin class with following re-registration."""
    try:
        admin.site.unregister(MODEL_TREE_CLASS)
    except NotRegistered:
        pass
    admin.site.register(MODEL_TREE_CLASS, _TREE_ADMIN())


def override_tree_admin(admin_class):
    """Sets a class that should be used instead of TreeAdmin
    to represent trees in the Admin interface.
    Note that the class must inherit from TreeAdmin.

    """
    global _TREE_ADMIN
    _TREE_ADMIN = lambda: admin_class
    _reregister_tree_admin()


def override_item_admin(admin_class):
    """Sets a class that should be used instead of TreeItemAdmin
    to represent tree items in the Admin interface.
    Note that the class must inherit from TreeItemAdmin.

    """
    global _ITEM_ADMIN
    _ITEM_ADMIN = lambda: admin_class
    _reregister_tree_admin()


class TreeItemAdmin(admin.ModelAdmin):

    exclude = ('tree', 'sort_order')
    fieldsets = (
        (_('Basic settings'), {
            'fields': ('parent', 'title', 'url',)
        }),
        (_('Access settings'), {
            'classes': ('collapse',),
            'fields': ('access_loggedin', 'access_guest', 'access_restricted', 'access_permissions', 'access_perm_type')
        }),
        (_('Display settings'), {
            'classes': ('collapse',),
            'fields': ('hidden', 'inmenu', 'inbreadcrumbs', 'insitetree')
        }),
        (_('Additional settings'), {
            'classes': ('collapse',),
            'fields': ('hint', 'description', 'alias', 'urlaspattern')
        }),
    )
    filter_horizontal = ('access_permissions',)
    change_form_template = 'admin/sitetree/treeitem/change_form.html'

    def response_add(self, request, obj, post_url_continue=None, **kwargs):
        """Redirects to the appropriate items' 'continue' page on item add.

        As we administer tree items within tree itself, we
        should make some changes to redirection process.

        """
        if post_url_continue is None:
            post_url_continue = '../item_%s/' % obj.pk
        return super(TreeItemAdmin, self).response_add(request, obj, post_url_continue)

    def response_change(self, request, obj, **kwargs):
        """Redirects to the appropriate items' 'add' page on item change.

        As we administer tree items within tree itself, we
        should make some changes to redirection process.

        """
        response = super(TreeItemAdmin, self).response_change(request, obj)
        if '_addanother' in request.POST:
            return HttpResponseRedirect('../item_add/')
        elif '_save' in request.POST:
            return HttpResponseRedirect('../')
        else:
            return HttpResponseRedirect('')

    def get_form(self, request, obj=None, **kwargs):
        """Returns modified form for TreeItem model.
        'Parent' field choices are built by sitetree itself.

        """
        if obj is not None and obj.parent is not None:
            self.previous_parent = obj.parent
            previous_parent_id = self.previous_parent.id
        else:
            previous_parent_id = None

        my_choice_field = TreeItemChoiceField(self.tree, initial=previous_parent_id)
        form = super(TreeItemAdmin, self).get_form(request, obj, **kwargs)
        my_choice_field.label = form.base_fields['parent'].label
        my_choice_field.help_text = form.base_fields['parent'].help_text
        # Replace 'parent' TreeItem field with new appropriate one
        form.base_fields['parent'] = my_choice_field

        # Try to resolve all currently registered url names including those in namespaces.
        if not getattr(self, 'known_url_names', False):
            self.known_url_names = []
            self.known_url_rules = []
            resolver = get_resolver(get_urlconf())
            for ns, (url_prefix, ns_resolver) in resolver.namespace_dict.items():
                if ns != 'admin':
                    self._stack_known_urls(ns_resolver.reverse_dict, ns)
            self._stack_known_urls(resolver.reverse_dict)
            self.known_url_rules = sorted(self.known_url_rules)

        form.known_url_names_hint = _(
            'You are seeing this warning because "URL as Pattern" option is active and pattern entered above '
            'seems to be invalid. Currently registered URL pattern names and parameters: ')
        form.known_url_names = self.known_url_names
        form.known_url_rules = self.known_url_rules
        return form

    def _stack_known_urls(self, reverse_dict, ns=None):
        for url_name, url_rules in reverse_dict.items():
            if isinstance(url_name, six.string_types):
                if ns is not None:
                    url_name = '%s:%s' % (ns, url_name)
                self.known_url_names.append(url_name)
                self.known_url_rules.append('<b>%s</b> %s' % (url_name, ' '.join(url_rules[0][0][1])))

    def get_tree(self, request, tree_id, item_id=None):
        """Fetches Tree for current or given TreeItem."""
        if tree_id is None:
            tree_id = self.get_object(request, item_id).tree_id
        self.tree = MODEL_TREE_CLASS._default_manager.get(pk=tree_id)
        self.tree.verbose_name_plural = self.tree._meta.verbose_name_plural
        self.tree.urls = _TREE_URLS
        return self.tree

    def item_add(self, request, tree_id):
        tree = self.get_tree(request, tree_id)
        return self.add_view(request, extra_context={'tree': tree})

    def item_edit(self, request, item_id, tree_id=None):
        tree = self.get_tree(request, tree_id, item_id)
        return self.change_view(request, item_id, extra_context={'tree': tree})

    def item_delete(self, request, item_id, tree_id=None):
        tree = self.get_tree(request, tree_id, item_id)
        return self.delete_view(request, item_id, extra_context={'tree': tree})

    def item_history(self, request, item_id, tree_id=None):
        tree = self.get_tree(request, tree_id, item_id)
        return self.history_view(request, item_id, extra_context={'tree': tree})

    def item_move(self, request, tree_id, item_id, direction):
        """Moves item up or down by swapping 'sort_order' field values of neighboring items."""
        current_item = MODEL_TREE_ITEM_CLASS._default_manager.get(pk=item_id)
        if direction == 'up':
            sort_order = 'sort_order'
        else:
            sort_order = '-sort_order'

        siblings = MODEL_TREE_ITEM_CLASS._default_manager.filter(
            parent=current_item.parent,
            tree=current_item.tree
        ).order_by(sort_order)

        previous_item = None
        for item in siblings:
            if item != current_item:
                previous_item = item
            else:
                break

        if previous_item is not None:
            current_item_sort_order = current_item.sort_order
            previous_item_sort_order = previous_item.sort_order

            current_item.sort_order = previous_item_sort_order
            previous_item.sort_order = current_item_sort_order

            current_item.save()
            previous_item.save()

        return HttpResponseRedirect('../../')

    def save_model(self, request, obj, form, change):
        """Saves TreeItem model under certain Tree.
        Handles item's parent assignment exception.

        """
        if change:
            # No, you're not allowed to make item parent of itself
            if obj.parent is not None and obj.parent.id == obj.id:
                obj.parent = self.previous_parent
                messages.warning(
                    request, _("Item's parent left unchanged. Item couldn't be parent to itself."), '', True)
        obj.tree = self.tree
        obj.save()


def redirects_handler(*args, **kwargs):
    """Fixes Admin contrib redirects compatibility problems
    introduced in Django 1.4 by url handling changes.

    """
    referer = args[0].META['HTTP_REFERER']
    shift = '../'

    if 'delete' in referer:
        # Weird enough 'delete' is not handled by TreeItemAdmin::response_change().
        shift += '../'
    elif 'history' in referer:
        if 'item_id' not in kwargs:
            # Encountered request from history page to return to tree layout page.
            shift += '../'

    return HttpResponseRedirect(referer + shift)


class TreeAdmin(admin.ModelAdmin):

    list_display = ('alias', 'title')
    list_display_links = ('title', 'alias')
    search_fields = ['title', 'alias']
    ordering = ['title', 'alias']
    actions = None
    change_form_template = 'admin/sitetree/tree/change_form.html'

    def __init__(self, *args, **kwargs):

        if SMUGGLER_INSTALLED:
            self.change_list_template = 'admin/sitetree/tree/change_list_.html'

        super(TreeAdmin, self).__init__(*args, **kwargs)
        self.tree_admin = _ITEM_ADMIN()(MODEL_TREE_ITEM_CLASS, admin.site)

    def get_urls(self):
        """Manages not only TreeAdmin URLs but also TreeItemAdmin URLs."""
        urls = super(TreeAdmin, self).get_urls()
        sitetree_urls = patterns('',
            # Trying to be nice and adopt url handling changes in Django 1.4, 1.5 Admin contrib.
            url(r'^/$', redirects_handler, name=get_tree_item_url_name('changelist')),
            url(r'^((?P<tree_id>\d+)/)?item_add/$',
                self.admin_site.admin_view(self.tree_admin.item_add), name=get_tree_item_url_name('add')),
            url(r'^(?P<tree_id>\d+)/item_(?P<item_id>\d+)/$',
                self.admin_site.admin_view(self.tree_admin.item_edit), name=get_tree_item_url_name('change')),
            url(r'^item_(?P<item_id>\d+)/$',
                self.admin_site.admin_view(self.tree_admin.item_edit), name=get_tree_item_url_name('change')),
            url(r'^((?P<tree_id>\d+)/)?item_(?P<item_id>\d+)/delete/$',
                self.admin_site.admin_view(self.tree_admin.item_delete), name=get_tree_item_url_name('delete')),
            url(r'^((?P<tree_id>\d+)/)?item_(?P<item_id>\d+)/history/$',
                self.admin_site.admin_view(self.tree_admin.item_history), name=get_tree_item_url_name('history')),
            url(r'^(?P<tree_id>\d+)/item_(?P<item_id>\d+)/move_(?P<direction>(up|down))/$',
                self.admin_site.admin_view(self.tree_admin.item_move), name=get_tree_item_url_name('move')),
        )

        if SMUGGLER_INSTALLED:
            sitetree_urls += (url(r'^dump_all/$', self.admin_site.admin_view(self.dump_view), name='sitetree_dump'),)

        return sitetree_urls + urls

    @classmethod
    def dump_view(cls, request):
        """Dumps sitetrees with items using django-smuggler.

        :param request:
        :return:
        """
        from smuggler.views import dump_to_response
        return dump_to_response(request, [MODEL_TREE, MODEL_TREE_ITEM], filename_prefix='sitetrees')


_reregister_tree_admin()
